### PR TITLE
Feat/delete user account

### LINKED ISSUE #349: [FEATURE] As a user I want to delete my account
**Is your feature request related to a problem? Please describe.**
Add an option to Profile Settings to delete the account.

**Describe the solution you'd like**
Delete Account Option

**Additional context**
The tricky part is the handling of memberships and teams. If the user is the only member of a team, delete the team with the user but if the account has multiple members:

Solution: Make the first admin that can be found to the new owner of the team.

If no admin is in the team, reject the delete account request with the reason that he needs to appoint an admin to the team first.

### PR DESCRIPTION
**Pull request by fdis111**
## As a user I want to delete my account

Fixes #349

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [ ] Enhancement (small improvements)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change adds a new database migration
- [ ] This change requires a documentation update

## How should this be tested?

- Login to the platform
- Click on the Settings option on the navigation bar 
- On the side navigation, click on profile below the account menu. 
- Click on the delete my account button 
- Confirm your deletion 

## Checklist

- [x] Read the [contributing guide](https://github.com/formbricks/formbricks/blob/main/CONTRIBUTING.md)
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand bits
- [x] Ran `pnpm build`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues
- [ ] Updated the Formbricks Docs if changes were necessary

### CODE REVIEW

**Comment by jobenjada** [comment]
Hey @fdis111 

I'm getting a build error:

grafik


Please fix and remember to always run `pnpm build` before shipping :)

**Comment by jobenjada** [comment]
Hey Francois! 

Overall the PR looks good, thanks for your contribution 🙏 

We have two questions:

1. **Filtering for admin:** In the issue description I mentioned that we will run into problems when you sign a team over to a person who is not an admin of the team.  Even though at this point all other users are admins, we plan to introduce more roles down the line. It would be great if you could check if there is an admin in the team and sign the team over to this user.

2. Assigning the admin role to the to-be-deleted user: Why do you make the user which will be deleted an admin before deletion?

3. Please run `git pull origin main` and resolve the merge conflict.

Thanks a lot!

File: apps/web/app/environments/[environmentId]/settings/profile/editProfile.tsx [hunk 1]
```diff
@@ -69,3 +75,86 @@ export function EditAvatar({ session }) {
...
+
+  return (
+    <DeleteDialog
+      open={open}
+      setOpen={setOpen}
+      deleteWhat="account"
+      onDelete={() => deleteAccount()}
+      text="Before you proceed with deleting your account, please be aware of the following consequences:"
+      isDeleting={deleting}
+      disabled={!(inputValue === session.user.email)}>
```
Comment (mattinannt) [inline apps/web/app/environments/[environmentId]/settings/profile/editProfile.tsx:115]: Using `!==` would be easier to read:
```
disabled={inputValue !== session.user.email}
```

**Review by mattinannt — CHANGES_REQUESTED** [review]
@fdis111 Thanks a lot for the PR 🤗 Generally great work, I made a few small suggestions for improvements that I marked in the code 😊💪.
The only critical part is that I haven't found the logic that deletes the team if the user is the only team member. This is needed to not have an orphaned team with no members left in the database after deleting the user.

File: apps/web/app/api/v1/users/me/route.ts [hunk 2]
```diff
@@ -37,3 +37,113 @@ export async function PUT(request: NextRequest) {
...
+        userId,
+        teamId,
+      },
+    },
+  });
+};
+
+// I created this type because I don't have access to prisma unums.
+// TODO find a way to get access to prisma enums
+type MembershipRole = "admin" | "owner";
```
Comment (mattinannt) [inline apps/web/app/api/v1/users/me/route.ts:61]: @fdis111 You can import `MembershipRole` from `@prisma/client`:

```import { MembershipRole } from "@prisma/client";```

File: apps/web/app/api/v1/users/me/route.ts [hunk 3]
```diff
@@ -37,3 +37,113 @@ export async function PUT(request: NextRequest) {
...
+
+      if (teamHasOnlyOneMember) {
+        await deleteTeam(teamId);
+      } else {
+        if (currentUserIsTeamOwner) {
+          if (teamHasAtLeastOneAdmin) {
+            const firstAdmin = teamAdminMemberships[0];
+            await updateUserMembership(teamId, firstAdmin.userId, "owner");
+          } else {
+            await deleteMembership(teamId, currentUser.id);
```
Comment (mattinannt) [inline apps/web/app/api/v1/users/me/route.ts:135]: In this case, the person deleting the account is the owner of the team, and there are other members who are not admins (viewers, editors,..).
Just deleting the owner's membership and leaving the team without an owner and without an admin seems like a bad option. In this case, please delete the whole team and we will update the delete info text.

File: apps/web/app/environments/[environmentId]/settings/profile/editProfile.tsx [hunk 4]
```diff
@@ -69,3 +75,86 @@ export function EditAvatar({ session }) {
...
+    <DeleteDialog
+      open={open}
+      setOpen={setOpen}
+      deleteWhat="account"
+      onDelete={() => deleteAccount()}
+      text="Before you proceed with deleting your account, please be aware of the following consequences:"
+      isDeleting={deleting}
+      disabled={inputValue !== session.user.email}>
+      <div className="py-5">
+        <p>
```
Comment (mattinannt) [inline apps/web/app/environments/[environmentId]/settings/profile/editProfile.tsx:117]: Please update to:
```jsx
          <p>
          Deleting your account will result in the permanent removal of all your personal information, saved preferences, and access to team data. If you are the owner of a team with other admins, the ownership of that team will be transferred to another admin.
        </p>
        <p className="py-5">
          Please note, however, that if you are the only member of a team or there is no other admin present, the team will be irreversibly deleted along with all associated data.
        </p>
```

**Review by mattinannt — CHANGES_REQUESTED** [review]
@fdis111 Thanks a lot for the changes 😊 There was one case where I think we need a slightly different behavior and I also made some small notes. Otherwise it looks great 🎉

**Review by mattinannt — CHANGES_REQUESTED** [review]
@fdis111 Thanks for the changes 💪 Everything works great 🎉
Please just add the last missing `await` and we are ready to merge the changes into main 🤗

**Pull request merged**